contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,290 @@
+    }
+
+    function undelegate(address delegator, uint256 shares) external onlyStakeHub returns (uint256) {
+        require(shares != 0, "ZERO_AMOUNT");
+        require(shares <= balanceOf(delegator), "INSUFFICIENT_BALANCE");
+
+        _lockedShares[delegator] += shares;
+
+        // calculate the BNB amount and update state
+        uint256 bnbAmount = _burnAndSync(delegator, shares);
+
+        // add to the queue
+        uint256 unlockTime = block.timestamp + IStakeHub(STAKE_HUB_ADDR).unbondPeriod();
+        UnbondRequest memory request = UnbondRequest({ shares: shares, bnbAmount: bnbAmount, unlockTime: unlockTime });
+        bytes32 hash = keccak256(abi.encodePacked(delegator, _useSequence(delegator)));
+        _unbondRequests[hash] = request;
+        _unbondRequestsQueue[delegator].pushBack(hash);
+
+        emit UnbondRequested(delegator, shares, bnbAmount, request.unlockTime);
+        return bnbAmount;
+    }
+
+    /**
+     * @dev Unbond immediately without adding to the queue.
+     * Only for redelegate process.
+     */
+    function unbond(address delegator, uint256 shares) external onlyStakeHub returns (uint256) {
+        require(shares != 0, "ZERO_AMOUNT");
+        require(shares <= balanceOf(delegator), "INSUFFICIENT_BALANCE");
+
+        // calculate the BNB amount and update state
+        uint256 bnbAmount = _burnAndSync(delegator, shares);
+
+        uint256 _gasLimit = IStakeHub(STAKE_HUB_ADDR).transferGasLimit();
+        (bool success,) = STAKE_HUB_ADDR.call{ gas: _gasLimit, value: bnbAmount }("");
+        require(success, "TRANSFER_FAILED");
+
+        emit Unbonded(delegator, shares, bnbAmount);
+        return bnbAmount;
+    }
+
+    function claim(address payable delegator, uint256 number) external onlyStakeHub nonReentrant returns (uint256) {
+        if (delegator == validator) {
+            require(!_freeze, "VALIDATOR_FROZEN");
+        }
+
+        // number == 0 means claim all
+        // number should not exceed the length of the queue
+        require(_unbondRequestsQueue[delegator].length() != 0, "NO_UNBOND_REQUEST");
+        number = (number == 0 || number > _unbondRequestsQueue[delegator].length())
+            ? _unbondRequestsQueue[delegator].length()
+            : number;
+
+        uint256 _totalShares;
+        uint256 _totalBnbAmount;
+        while (number != 0) {
+            bytes32 hash = _unbondRequestsQueue[delegator].front();
+            UnbondRequest memory request = _unbondRequests[hash];
+            if (block.timestamp < request.unlockTime) {
+                break;
+            }
+
+            _totalShares += request.shares;
+            _totalBnbAmount += request.bnbAmount;
+
+            // remove from the queue
+            _unbondRequestsQueue[delegator].popFront();
+            delete _unbondRequests[hash];
+
+            --number;
+        }
+        require(_totalShares != 0, "NO_CLAIMABLE_UNBOND_REQUEST");
+        _lockedShares[delegator] -= _totalShares;
```

Comment:
I think `_lockedShares` is uselss, but it helps to secure in some way. I prefer `_lockedBNB` than share here.

Reply:
Because when a delegator request for unbond, his shares will be burned immediately but cannot get the BNB at once. So I add this to help delegator to know how many shares "burned but not cased" he has